src/runtime_src/core/pcie/linux/device_linux.cpp

```diff
@@ -451,6 +451,111 @@ struct hotplug_offline
   }
 };
 
+struct clk_scaling_info
+{
+  using result_type = query::clk_scaling_info::result_type;
+  using data_type = query::clk_scaling_info::data_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type)
+  {
+    auto pdev = get_pcidev(device);
+
+    std::vector<std::string> stats;
+    std::string errmsg;
+    result_type ctStats;
+
+    pdev->sysfs_get("xgq_vmr", "clk_scaling_stat_raw", errmsg, stats);
+    if (!errmsg.empty()) {
+      // Backward compatibilty check.
+      // Read XMC sysfs nodes
+      data_type data = { 0 };
+	  uint32_t uint_op = 0;
+
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_enabled", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.enable = uint_op ? true : false;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_support", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.support = uint_op ? true : false;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_critical_power_threshold", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.pwr_shutdown_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_critical_temp_threshold", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.temp_shutdown_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_power_limit", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.pwr_scaling_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_temp_limit", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.temp_scaling_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_temp_override", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.temp_scaling_ovrd_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_power_override", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.pwr_scaling_ovrd_limit = uint_op;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_power_override_en", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.pwr_scaling_ovrd_enable = uint_op ? true : false;
+      pdev->sysfs_get<uint32_t>("xmc", "scaling_threshold_temp_override_en", errmsg, uint_op, EINVAL);
+      if (errmsg.empty())
+        data.temp_scaling_ovrd_enable = uint_op ? true : false;
+      ctStats.push_back(data);
+      return ctStats;
+	}
+
+    data_type data = { 0 };
+    int id = 0;
+    //parse one line at a time
+    // The clk_scaling_stat_raw is printing in formatted string of each line
+    // Format: "%s:%u"
+    // Using colon ":" as separator.
+    for (auto& stat_raw : stats) {
+      std::vector<std::string> stat;
+      boost::split(stat, stat_raw, boost::is_any_of(":"));
+      uint16_t value = std::stoi(std::string(stat.at(1)));
+      switch (id) {
```

Comment:
This switch statement isn't very scalable and it assumes that the stat array is order.  Meaning the first item is support and the last is temp_scaling_ovrd_enable.  

Minimally, there should be a helper function:
```
uint16_t get_value(const std::string & keyValue)
{
      std::vector<std::string> stat;
      boost::split(stat, keyValue, boost::is_any_of(":"));
     if (stat.size() != 2) {
        const auto errMsg = boost::format("Error: KeyValue pair doesn't meet expected format '<key>:<value>': '%s'") % keyValue;
        throw std::runtime_error(errMsg.str())
      return = std::stoi(std::string(stat.at(1)));
}
```
And it would be to initialize each of the variables:
```
data.support = get_value(stats[0]) ? true : false;
data.enable = get_value(stats[1]) ? true : false;
data.pwr_shutdown_limit = get_value(stats[2]);
```

Of course, it would be better if the key was use to data drive the variables.  This can be done without the need of a switch statement.